File: 62-unique-paths/62-unique-paths.cpp
class Solution {
public:
    //int ans = 0;
    int bfs(int row, int column, int m, int n, vector<vector<int>> &dp){
        if(row == 0 and column == 0){
            return 1;
        }
        else if(row < 0 or column < 0) return 0;
        else if(dp[row][column] != -1) return dp[row][column];
        
        //since we're going bottom up dp
        int up = bfs(row-1, column, m, n, dp);
        int left = bfs(row, column-1, m, n, dp);
        
        return dp[row][column] = up + left;
         
    }
    
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m+1, vector<int>(n+1, -1));
        return bfs(m-1,n-1,m,n, dp);
    }
};

File: 62-unique-paths/62-unique-paths.py
class Solution:
    
    def uniquePaths(self, m: int, n: int) -> int:
        #dp = [[-1]*n]*m
        
        dp = {}
                
        def helper(row, col, dp):
            print(dp)
            #this is one way to reach the dest
            if row == 0 and col == 0: return 1

            if row < 0 or col < 0: return 0

            if (row,col) in dp: return dp[(row,col)]

            left = helper(row, col-1, dp)
            up = helper(row-1, col, dp)
            dp[(row,col)] = left+up
            return dp[(row,col)]
        
        return helper(m-1, n-1, dp)